File: android/csharp/UI/ConfigChooser.cs
namespace Carto.Ui {
    using Javax.Microedition.Khronos.Egl;
    using Android.Util;

    // EGL configuration chooser for MapView.
    internal class ConfigChooser : Java.Lang.Object, Android.Opengl.GLSurfaceView.IEGLConfigChooser {

        private static int[][] ATTRIBUTE_TABLE = new int[][] {
            // 8-8-8-8-bit color, 8-bit stencil, 24-bit z buffer. Should work on most devices.
            new int[] { EGL10.EglRedSize, 8, EGL10.EglGreenSize, 8, EGL10.EglBlueSize, 8, EGL10.EglAlphaSize, 8, EGL10.EglDepthSize, 24, EGL10.EglStencilSize, 8, EGL10.EglRenderableType, 4, EGL10.EglNone },
            // 8-8-8-8-bit color, 8-bit stencil, 16-bit z buffer. Better than 5-6-5/16 bit, should also fix problems on some obscure devices.
            new int[] { EGL10.EglRedSize, 8, EGL10.EglGreenSize, 8, EGL10.EglBlueSize, 8, EGL10.EglAlphaSize, 8, EGL10.EglDepthSize, 16, EGL10.EglStencilSize, 8, EGL10.EglRenderableType, 4, EGL10.EglNone },
            // 5-6-5-bit color, 24-bit z buffer. Should work on most devices.
            new int[] { EGL10.EglRedSize, 5, EGL10.EglGreenSize, 6, EGL10.EglBlueSize, 5, EGL10.EglDepthSize, 24, EGL10.EglRenderableType, 4, EGL10.EglNone },
            // 5-6-5-bit color, 16-bit z buffer. Fallback for original Tegra devices. 
            new int[] { EGL10.EglRedSize, 5, EGL10.EglGreenSize, 6, EGL10.EglBlueSize, 5, EGL10.EglDepthSize, 16, EGL10.EglRenderableType, 4, EGL10.EglNone },
            // 5-6-5-bit color, unspecified z/stencil buffer.
            new int[] { EGL10.EglRedSize, 5, EGL10.EglGreenSize, 6, EGL10.EglBlueSize, 5, EGL10.EglRenderableType, 4, EGL10.EglNone },
        };

        public Javax.Microedition.Khronos.Egl.EGLConfig ChooseConfig(Javax.Microedition.Khronos.Egl.IEGL10 egl, Javax.Microedition.Khronos.Egl.EGLDisplay eglDisplay) {
            Carto.Utils.Log.Debug("ConfigChooser.ChooseConfig: Model: " + Android.OS.Build.Model + ", board: " + Android.OS.Build.Board + ", product: " + Android.OS.Build.Product);

            for (int i = 0; i < ATTRIBUTE_TABLE.Length; i++) {
                int[] numConfigs = new int[] { 0 };
                EGLConfig[] configs = new EGLConfig[1];
                if (egl.EglChooseConfig(eglDisplay, ATTRIBUTE_TABLE[i], configs, 1, numConfigs)) {
                    if (numConfigs[0] > 0) {
                        Carto.Utils.Log.Debug("ConfigChooser.ChooseConfig: Selected display configuration: " + i);
                        return configs[0];
                    }
                }
            }
            return null;
        }
    }
}

File: android/java/com/carto/ui/ConfigChooser.java
package com.carto.ui;

import javax.microedition.khronos.egl.EGL10;
import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.egl.EGLDisplay;

import android.opengl.GLSurfaceView;

/**
 * EGL configuration chooser for MapView and TextureMapView.
 * It is not intended for public usage.
 * @hidden
 */
public class ConfigChooser implements GLSurfaceView.EGLConfigChooser {
    
    private static int[][] ATTRIBUTE_TABLE = new int[][] {
        // 8-8-8-8-bit color, 8-bit stencil, 24-bit z buffer. Should work on most devices.
        new int[] { EGL10.EGL_RED_SIZE, 8, EGL10.EGL_GREEN_SIZE, 8, EGL10.EGL_BLUE_SIZE, 8, EGL10.EGL_ALPHA_SIZE, 8, EGL10.EGL_DEPTH_SIZE, 24, EGL10.EGL_STENCIL_SIZE, 8, EGL10.EGL_RENDERABLE_TYPE, 4, EGL10.EGL_NONE },
        // 8-8-8-8-bit color, 8-bit stencil, 16-bit z buffer. Better than 5-6-5/16 bit, should also fix problems on some obscure devices.
        new int[] { EGL10.EGL_RED_SIZE, 8, EGL10.EGL_GREEN_SIZE, 8, EGL10.EGL_BLUE_SIZE, 8, EGL10.EGL_ALPHA_SIZE, 8, EGL10.EGL_DEPTH_SIZE, 16, EGL10.EGL_STENCIL_SIZE, 8, EGL10.EGL_RENDERABLE_TYPE, 4, EGL10.EGL_NONE },
        // 5-6-5-bit color, 24-bit z buffer, unspecified stencil. Should work on most devices.
        new int[] { EGL10.EGL_RED_SIZE, 5, EGL10.EGL_GREEN_SIZE, 6, EGL10.EGL_BLUE_SIZE, 5, EGL10.EGL_DEPTH_SIZE, 24, EGL10.EGL_RENDERABLE_TYPE, 4, EGL10.EGL_NONE },
        // 5-6-5-bit color, 16-bit z buffer, unspecified stencil. Fallback for original Tegra devices. 
        new int[] { EGL10.EGL_RED_SIZE, 5, EGL10.EGL_GREEN_SIZE, 6, EGL10.EGL_BLUE_SIZE, 5, EGL10.EGL_DEPTH_SIZE, 16, EGL10.EGL_RENDERABLE_TYPE, 4, EGL10.EGL_NONE },
        // 5-6-5-bit color, unspecified z/stencil buffer.
        new int[] { EGL10.EGL_RED_SIZE, 5, EGL10.EGL_GREEN_SIZE, 6, EGL10.EGL_BLUE_SIZE, 5, EGL10.EGL_RENDERABLE_TYPE, 4, EGL10.EGL_NONE },
    };

    @Override
    public EGLConfig chooseConfig(EGL10 egl, EGLDisplay eglDisplay) {
        com.carto.utils.Log.debug("ConfigChooser.chooseConfig: Model: " + android.os.Build.MODEL + ", board: " + android.os.Build.BOARD + ", product: " + android.os.Build.PRODUCT);

        for (int i = 0; i < ATTRIBUTE_TABLE.length; i++) {
            int[] numConfigs = new int[] { 0 };
            EGLConfig[] configs = new EGLConfig[1];
            if (egl.eglChooseConfig(eglDisplay, ATTRIBUTE_TABLE[i], configs, 1, numConfigs)) {
                if (numConfigs[0] > 0) {
                    com.carto.utils.Log.debug("ConfigChooser.chooseConfig: Selected display configuration: " + i);
                    return configs[0];
                }
            }
        }
        throw new IllegalArgumentException("Failed to choose EGLConfig!");
    }
}
